Reformat Contributing (#6251)
TownyAdvanced/Towny: .github/CONTRIBUTING.MD

# Contributing to Towny

> Thanks for deciding to contribute to Towny! We really appreciate contributions, but to maintain everything cool around here, we need to enforce some guidelines to keep the project as efficient as possible. :shipit:

Reviewer: ```suggestion
Thanks for deciding to contribute to Towny! We really appreciate contributions, but to keep everything running smoothly and get contributions accepted in a timely manner, we need to enforce some guidelines. :shipit:
```

Contributions are welcomed via GitHub Pull Requests.
You must fork this repository and then open it via your IDE.   
You can always ask for help on your contribution in the [Towny Discord](https://discord.gg/gnpVs5m).    

## Before contributing
- Make sure to read the [Towny license](https://github.com/TownyAdvanced/Towny/blob/master/LICENSE.md).
- By contributing, you agree to distribute your changes under the Towny license and all its terms.

If you do not agree with the terms of our license, do not contribute to the Towny project.

# About pull requests

> 1. Keep your commit messages self-explainatory, the description should be a brief explanation and include an issue reference if applicable.

Reviewer: ```suggestion
1. Keep your commit messages self-explanatory. Pull requests will have their Commits squashed upon merge but the commit messages will be summarized.
```

> 2. Pull requests should be straight-forward and focus only on one feature. Changed files should only show a difference where code has been actually modified.

Reviewer: ```suggestion
2. Pull Requests forms should be filled out to the best of your abilities: descriptions, changes and issue references where applicable.
```

> 3. Test your changes before sending them in! You should look for bugs, test configuration values, commands, and anything you've touched!

Reviewer: ```suggestion
3. Your pull request should address a single topic, or be as closed in scope as you can reasonably make it. Small pull requests are easier to review and more predictable to test.
4. Test your changes before sending them in! You should look for bugs, test configuration values, commands, and anything you've touched!
```

# About Java code
> Towny uses an [.editorconfig](https://editorconfig.org) file. IDEs that have [native EditorConfig support](https://editorconfig.org/#pre-installed) will automatically follow these rules.
    
1. You must use tabs for indentation. 
2. Keep your environment in Java 16-17. The code will automatically be converted into code compatible with Java 8 by [Jabel](https://github.com/bsideup/jabel).[^jabel-warning]
3. Avoid using wildcard imports.
4. Do not format entire class files.   
   - All IDEs are different, we do this to keep consistency in the codebase.
   - This also keeps code differences in GitHub clean and simple.

> 5. Do not use NMS (`net.minecraft.server`)[^nms], or reflection. We aim at only using the Spigot API.
> 6. Do not directly interact with the Minecraft protocol.
>    - This includes the usage of ProtocolLib or Netty.

Reviewer: ```suggestion
5. Do not make unnecessary changes.
    - This includes whitespace changes.
    - We want to have as clean a diff as possible.
6. Do not use NMS (`net.minecraft.server`)[^nms], or reflection. We aim to use only the Spigot/Bukkit API, with some additional systems like PaperLib.
7. Do not directly interact with the Minecraft protocol.
   - This includes the usage of ProtocolLib or Netty.
```

1. Libraries need to be approved by the maintainer before being merged. We recommend you ask before commiting.     
2. Prevent `ClassNotFound` exceptions.
   - Do not reference a plugin in a loaded class, unless that plugin is enabled.
3. Do not add plugin support directly into Towny **unless** it can benefit the project (such as economy/permission bridges).
   - For Towny support, you should be using the [`TownyAPI`](https://github.com/TownyAdvanced/Towny/wiki/TownyAPI) instead.
   - Even if you believe it can benefit the project, please ask the maintainer first.
4. Do not embed dependencies directly into Towny.
   - All dependencies should use the `provided` scope instead. This means that the dependency will be provided by the plugin JAR (e.g. Vault), or the server JAR (e.g. Adventure).
5. Do not add closed source dependencies.
   - All dependencies should have open and auditable source code.

# About translation
All translation on the core Towny plugin is done via [Crowdin](https://crowdin.com/project/townyadvanced).

> Strings are formatted using [Java's `java.util.Formatter`](https://docs.oracle.com/en/java/javase/18/docs/api/java.base/java/util/Formatter.html) allowing specification of variable order, and more.

Reviewer: ```suggestion
Strings are formatted using [Java's `java.util.Formatter`](https://docs.oracle.com/en/java/javase/18/docs/api/java.base/java/util/Formatter.html) allowing specification of variable order, and more. If you find that a line has multiple variables which would be in a different order in your language, you can change them to `%Variable number$s` to change their order. 
An example of this: `%4$s` will **always use the 4th variable**, no matter where it is placed on the line.
```

# Contributing to Towny
Thanks for deciding to contribute to Towny! We really appreciate contributions, but to keep everything running smoothly and get contributions merged in a timely manner, we need to enforce some guidelines. :shipit:

> Contributions are welcomed via GitHub Pull Requests.

Reviewer: ```suggestion
Contributions are welcome via GitHub Pull Requests.
```

You must fork this repository and then open it via your IDE.   
You can always ask for help on your contribution in the [Towny Discord](https://discord.gg/gnpVs5m).    

## Before contributing
- Make sure to read the [Towny license](https://github.com/TownyAdvanced/Towny/blob/master/LICENSE.md).
- By opening an pull request, you:
    - represent that you have the right to waive your changes' copyright and any related rights.

>     - acknowledge that the TownyAdvanced organization is allowed to use or modify your changes in any way it sees fit and in accordance with the [Towny license](https://github.com/TownyAdvanced/Towny/blob/master/LICENSE.md)

Reviewer: This has to be redone so that there is no change from the original wording.

Author: Understood!

> If you do not agree with the terms of our license, **do not contribute to the Towny project**.

Reviewer: ```suggestion
If you do not agree with the terms of our license, do not contribute to the Towny project.
```

# About pull requests

> > Pull requests will have their commits squashed upon merge, commit messages will be summarized.

Reviewer: ```suggestion
> Pull requests will have their commits squashed upon merge, in doing so your various Commit Messages will be amalgamated into one Commit Message. Consider this while you create your individual commits.
```

1. Keep your commit messages self-explanatory. Keep them short but informative!

> 2. Pull requests should address a single topic, or be as closed in scope as you can reasonably make it. Small pull requests are easier to review and thus, can result in an faster merge.

Reviewer: ```suggestion
2. Pull requests should address a single topic, or be as closed in scope as you can reasonably make it. Small pull requests are easier to review and thus, can result in a faster merge.
```
